prelude.links

```diff
@@ -463,19 +463,75 @@ fun asList(t) server {
 # @param {[(String, String)]} attributes List of attribute-value-pairs
 # @param {Xml} children Child XML nodes
 # @returns {Xml}
-sig makeXml : (String, [(String, String)], Xml) ~> Xml
-fun makeXml (name, attributes, children) {
-  var attributes = fold_left(fun (prev, curr) {
-    var (attr, value) = curr;
-    prev ^^ " " ^^ attr ^^ "=\"" ^^ value ^^ "\""
-  }, "", attributes);
+sig makeXml : (String, [(String, String)], Xml) ~> XmlItem
+fun makeXml (name, attrs, children) {
+  hd(variantToXml([ Node(name, map(fun (attr) { var (n, v) = attr; Attr(n, v) }, attrs) ++ xmlToVariant(children)) ]))
+}
+
+# Appends a XML forest to an XmlItems children
+# @param {XmlItem} item
+# @param {Xml} child
+# @returns {XmlItem}
+sig appendChild : (XmlItem, Xml) ~> XmlItem
+fun appendChild(item, child) {
+  switch (xmlToVariant([ item ])) {
+    case [ Node(n, c) ] ->
+      hd(variantToXml([ Node(n, c ++ xmlToVariant(child)) ]))
+    case _ -> error("Cannot append to " ^^ show(item))
+  }
+}
+
+# Sets or overrides an attribute of a XmlItem
+# @param {XmlItem} item
+# @param {String} name Name of the attribute
+# @param {String} value Value the attribute is set to
+# @returns {XmlItem} Updated item
+sig setAttribute : (XmlItem, String, String) ~> XmlItem
+fun setAttribute (item, name, value) {
+  fun helper (i, o) {
+    switch (i) {
+      case [ ] -> reverse(Attr(name, value) :: o)
+      case Attr(n, v)::t -> if (n == name) {
+        reverse(Attr(n, value) :: o)
+      } else {
+        helper(t, Attr(n, v) :: o)
+      }
+      case other::t -> helper(t, other :: o)
+    }
+  }
 
-  var startTag = "<" ^^ name ^^ attributes ^^ ">";
-  var endTag = "</" ^^ name ^^ ">";
+  switch (xmlToVariant([ item ])) {
+    case [ Node(n, c) ] -> hd(variantToXml([ Node(n, helper(c, [ ])) ]))
+    case _ -> error("Cannot set attribute of " ^^ show(item))
+  }
+}
 
-  parseXml(startTag ^^ show(children) ^^ endTag)
+# Adds child to the front of an XmlItems children
+# Note: It is possible to add multiple children by passing an XML forest 
+# @param {XmlItem} item
+# @param {Xml} child or children
+# @returns {XmlItem}
+sig prependChild : (XmlItem, Xml) ~> XmlItem
+fun prependChild (item, child) {
+  switch (xmlToVariant([ item ])) {
+    case [ Node(n, c) ] -> hd(variantToXml([ Node(n, xmlToVariant(child) ++ c ) ]))
+    case _ -> error("Cannot prepend child in " ^^ show(item))
+  }  
+}
+
+# Replaces all children of an XmlItem
+# @param {XmlItem} item
+# @param {Xml} newChildren
+# @returns {XmlItem} item with new children
+sig replaceChildren : (XmlItem, Xml) ~> XmlItem
+fun replaceChildren (item, newChildren) {
+  switch (xmlToVariant([ item ])) {
+    case [ Node(n, c) ] -> hd(variantToXml([ Node(n, xmlToVariant(newChildren)) ]))
+    case _ -> error("Cannot replace children of " ^^ show(item))
+  }
```

Comment:
These operations seem fine, but are they needed given that we already have direct update operations to the DOM tree?